nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,337 @@
+        self.indexers.clear()
+
+    async def index_message_soon(self, msg: Msg):
+        work = WorkUnit.from_msg(msg)
+        await work.mpu.start()
+
+        logger.info(
+            f"Index message for shard {work.shard_id} is being enqueued",
+            extra={
+                "shard": work.shard_id,
+                "subject": msg.subject,
+                "reply": msg.reply,
+                "seqid": work.seqid,
+                "storage_key": work.index_message.storage_key,
+            },
+        )
+
+        await self.push_work_to_indexer(work)
+        CONCURRENT_INDEXERS_COUNT.set(
+            len(self.indexers), labels=dict(node=work.node_id)
+        )
+
+    async def push_work_to_indexer(self, work: WorkUnit):
+        exists = work.shard_id in self.indexers
+        if exists:
+            indexer, _ = self.indexers[work.shard_id]
+            await indexer.index_soon(work)
+        else:
+            indexer = PriorityIndexer(writer=self.writer, storage=self.storage)
+            await indexer.index_soon(work)
+            task = asyncio.create_task(indexer.work_until_finish())
+            task.add_done_callback(
+                partial(
+                    ConcurrentShardIndexer._indexer_done_callback, self, work.shard_id
+                )
+            )
+            self.indexers[work.shard_id] = (indexer, task)
+
+    def _indexer_done_callback(self, shard_id: str, task: asyncio.Task):
+        # autoremove indexer when finished
+        self.indexers.pop(shard_id)
+
+        # propagate errors
+        if task.exception() is not None:
+            raise task.exception()  # type: ignore
+
+
+class PriorityIndexer:
+    def __init__(self, writer: Writer, storage: Storage):
+        self.writer = writer
+        self.storage = storage
+        self.work_queue: asyncio.PriorityQueue = asyncio.PriorityQueue()
+        self.working = False
+
+    async def index_soon(self, work: WorkUnit):
+        await self.work_queue.put(work)
+
+    async def work_until_finish(self):
+        self.working = True
+        try:
+            while not self.work_queue.empty():
+                work = await self.work_queue.get()
+                await self._do_work(work)
+        except Exception as exc:
+            # if an exception occurred, we can't longer ensure proper ordering
+            # for this queue. We flush the queue and wait for the messages to be
+            # redelivered
+            event_id = capture_exception(exc)
+            logger.error(
+                "An error happened on indexer, all messages for this queue will be flushed. "
+                f"Check sentry for more details. Event id: {event_id}",
+                extra={
+                    "seqid": work.seqid,
+                    "shard": work.index_message.shard,
+                    "storage_key": work.index_message.storage_key,
+                },
+                exc_info=exc,
+            )
+            while not self.work_queue.empty():
```

Comment:
Alright so we're just going to wait for it to redeliver and not bubble an error into hard failure.

I wonder if we should also hard fail somehow but am okay starting here

Reply:
cause an exit with code 1 to show the process failed

Reply:
I see. I don't have an strong opinion on this. If you prefer some approach, we can go for it